codalab/worker/worker.py

```diff
@@ -269,15 +269,18 @@ def start(self):
             self.dependency_manager.start()
         while not self.terminate:
             try:
-                self.process_runs()
-                self.save_state()
                 self.checkin()
-                self.check_termination()
-                self.save_state()
-                if self.check_idle_stop() or self.check_num_runs_stop():
-                    self.terminate = True
-                else:
-                    time.sleep(self.checkin_frequency_seconds)
+                last_checkin = time.time()
+                # Process runs until it's time for the next checkin.
+                while not self.terminate and (
+                    time.time() - last_checkin <= self.checkin_frequency_seconds
```

Comment:
Can't we just get rid of this check then?

Reply:
Not sure what you mean? We can't get rid of the checkin_frequency_seconds check because we still want to check in periodically with the rest server.

Reply:
In which method does the check-in happen?

Reply:
Not sure, I'm guessing hundreds?

Reply:
oh I see, so the inner loop is pretty fast. That makes sense. Thanks!